### ISSUE TITLE
Rule suggestion: add 'no istanceof' rule

### ISSUE THREAD

**Issue by AsherBarak** [palantir/tslint#2943]
The way `instanceof` works presents a potential pitfall for ts developers. An instance passing ts type check may (or may not) fail `instanceof` check depending on the way it was constructed:
```ts
class A {
    a: string;
}

let a1: A = new A();
let a2: A = { a: 'test' };

console.log(a1 instanceof A); // true
console.log(a2 instanceof A); // false
```
Rule should flag usages of `instanceof` as unsafe.

see Microsoft/TypeScript#16693

**Comment by adidahiya** [comment]
IMO, in your example, `a2` really isn't an instance of `A` since `A.prototype` is not on the prototype chain. I don't see anything wrong with your code sample. I agree with the linked discussion that this check is better off as a lint rule rather than a compiler flag, but I'm hesitant to add it to TSLint's core rules unless there's sufficient demand from many users. Until then, feel free to implement it as a custom rule.

**Comment by JoshuaKGoldberg** [comment]
> hesitant to add it to TSLint's core rules unless there's sufficient demand from many users

Looks like there hasn't really been much interest here. As was mentioned here and in TypeScript-land, `instanceof` intentionally works only for classes. Using it on interface literals is a symptom of misusing the difference between classes and literals.

Closing for now, but if you feel strongly this should be a part of TSLint core, please do comment here!
